homeassistant/components/ring/sensor.py

```diff
@@ -71,171 +73,181 @@ def __init__(
-        else:
-            for entry in history_data:
-                if entry["kind"] == kind:
-                    found = entry
-                    break
-
-        if not found:
-            return
-
-        self._latest_event = found
+
+        self._device = self._get_coordinator_data().get_device(
+            self._device.device_api_id
+        )
+        # History values can drop off the last 10 events so only update
+        # the value if it's not None
+        if native_value := self.entity_description.value_fn(self._device):
+            self._attr_native_value = native_value
+        if self.entity_description.extra_state_attributes_fn and (
+            extra_attrs := self.entity_description.extra_state_attributes_fn(
+                self._device
+            )
+        ):
+            self._attr_extra_state_attributes = extra_attrs
         super()._handle_coordinator_update()
 
-    @property
-    def native_value(self):
-        """Return the state of the sensor."""
-        if self._latest_event is None:
-            return None
 
-        return self._latest_event["created_at"]
+def _get_last_event(
+    history_data: list[dict[str, Any]], kind: RingEventKind | None
+) -> dict[str, Any] | None:
+    if not history_data:
+        return None
+    if kind is None:
+        return history_data[0]
+    for entry in history_data:
+        if entry["kind"] == kind.value:
+            return entry
+    return None
+
+
+def _get_last_event_attrs(
+    history_data: list[dict[str, Any]], kind: RingEventKind | None
+) -> dict[str, Any] | None:
+    if last_event := _get_last_event(history_data, kind):
+        return {
+            "created_at": last_event.get("created_at"),
+            "answered": last_event.get("answered"),
+            "recording_status": last_event.get("recording", {}).get("status"),
+            "category": last_event.get("kind"),
+        }
+    return None
 
-    @property
-    def extra_state_attributes(self):
-        """Return the state attributes."""
-        attrs = super().extra_state_attributes
 
-        if self._latest_event:
-            attrs["created_at"] = self._latest_event["created_at"]
-            attrs["answered"] = self._latest_event["answered"]
-            attrs["recording_status"] = self._latest_event["recording"]["status"]
-            attrs["category"] = self._latest_event["kind"]
+@dataclass(frozen=True, kw_only=True)
+class RingSensorEntityDescription(SensorEntityDescription):
+    """Describes Ring sensor entity."""
 
-        return attrs
+    kind: RingEventKind | None = None
+    value_fn: Callable[[RingGeneric], StateType]
+    exists_fn: Callable[[RingGeneric], bool] = lambda _: True
+    extra_state_attributes_fn: (
+        Callable[[RingGeneric], MutableMapping[str, Any] | None] | None
+    ) = None
 
 
 @dataclass(frozen=True, kw_only=True)
-class RingSensorEntityDescription(SensorEntityDescription):
+class RingDoorbellSensorEntityDescription(RingSensorEntityDescription):
```

Comment:
One could make the argument that `RingDoorbellSensorEntityDescription` and `RingOtherSensorEntityDescription` are not really needed. The same can be archived slightly differently.

For the `value_fn`, make `RingSensorEntityDescription` a generic like so
```py
from typing_extensions import TypeVar  # import form "typing_extensions"!
_RingGenericT = TypeVar("_RingGenericT", bound=RingGeneric, default=RingGeneric)
class RingSensorEntityDescription(SensorEntityDescription, Generic[_RingGenericT]):

    value_fn: Callable[[_RingGenericT], StateType]
    exists_fn: Callable[[RingGeneric], bool]  # Note this stays "RingGeneric"!
```
As for the `exist_fn` overwrites. Those can be inlined in the description definitions.

Subsequently those are changed as follows:
```diff
-RingDoorbellSensorEntityDescription
+RingSensorEntityDescription[RingDoorBell]

-RingOtherSensorEntityDescription
+RingSensorEntityDescription[RingOther]
```

Reply:
Resolving here based on https://github.com/home-assistant/core/pull/115276#discussion_r1559252113 and later suggestions